zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ListenerContainer.java

```diff
@@ -0,0 +1,186 @@
+  private final ReentrantLock _lock = new ReentrantLock(true);
+  private final Map<String, Set<T>> _persistentListener = new ConcurrentHashMap<>();
+  private final Map<String, Set<T>> _onetimeListener = new ConcurrentHashMap<>();
+
+  /**
+   * Add listener to the container with specified key.
+   * @param key the key to register to
+   * @param listener the listener to register
+   * @param persistListener true if the listener is persistent
+   */
+  public void addListener(String key, T listener, boolean persistListener) {
+    addListener(key, listener, persistListener ? _persistentListener : _onetimeListener);
+  }
+
+  /**
+   * Remove listener from the container.
+   * This operation removes both one-time and persistent listener from the specified key.
+   * @param key the key to remove
+   * @param listener the listener to remove
+   */
+  public void removeListener(String key, T listener) {
+    _lock.lock();
+    try {
+      removeFromListenerMap(key, listener, _persistentListener);
+      removeFromListenerMap(key, listener, _onetimeListener);
+    } finally {
+      _lock.unlock();
+    }
+  }
+
+  /**
+   * Remove listener from the container.
+   * @param key the key to remove
+   * @param listener the listener to remove
+   * @param persistListener true if remove the persistent listener, otherwise remove the one-time listener
+   */
+  public void removeListener(String key, T listener, boolean persistListener) {
+    _lock.lock();
+    try {
+      if (persistListener) {
+        removeFromListenerMap(key, listener, _persistentListener);
+      } else {
+        removeFromListenerMap(key, listener, _onetimeListener);
+      }
+    } finally {
+      _lock.unlock();
+    }
+  }
+
+  /**
+   * Remove listeners from the container.
+   * @param key the key to remove
+   * @param listeners the listeners to remove
+   * @param persistListener true if remove the persistent listener, otherwise remove the one-time listener
+   */
+  public void removeListeners(String key, Collection<T> listeners, boolean persistListener) {
+    _lock.lock();
+    try {
+      if (persistListener) {
+        removeFromListenerMap(key, listeners, _persistentListener);
+      } else {
+        removeFromListenerMap(key, listeners, _onetimeListener);
+      }
+    } finally {
+      _lock.unlock();
+    }
+  }
+
+  /**
+   * Consume the listeners registered to the given key. The given consumer is applied to all one-time and persistent
+   * listeners with the given key.
+   * The one-time listeners are removed after this operation, as a result, this operation is NOT idempotent.
+   * @param key the key of the listeners
+   * @param consumer the consuming action for the listeners
+   */
+  public void consumeListeners(String key, Consumer<? super T> consumer) {
+    Set<T> onetimeListeners;
+    Set<T> persistentListeners;
+    _lock.lock();
+    try {
+      // remove one-time listeners
+      onetimeListeners = _onetimeListener.remove(key);
```

Comment:
Please correct me if I'm wrong but I believe the remove method in ConcurrentHashMap are atomic in itself so not sure if it is necessary to lock beforehand.

Also should we check if key is null? Given underlying data structure of a ConcurrentHashMap is a HashTable, attempting to remove a null key would throw a NullPointerException which might not be the desired effect. Just a thought

Reply:
Good question, I think there is a subtle race condition between `consume` and `add`. When thread A adds a one-time listener T, while thread B consume it as in line 122, if the timing is unfortunate, `onetimeListener` may be getting the snapshot before A completes, which doesn't contain T. However, because the entry together with the COWS instance is removed, the new listener T is lost.
(The unit test fails if we remove the lock)

Reply:
for deterministic test behavior, i think, lock will help. but we code for correctness of the system.

If you are going to use lock explicitly, then you can use regular HashMap. We are paying the cost twice as ConcurrentHashMap implementation has lock per segment:
https://github.com/frohoff/jdk8u-jdk/blob/master/src/share/classes/java/util/concurrent/ConcurrentHashMap.java#L1370

Reply:
Thanks @desaikomal for the pointer. Short answer on this, we need the lock for correctness, and it's the current impl.
The use of CopyOnWriteArraySet as entry in ConcurrentHashMap requires a lock to make the overall structure thread-safe. Otherwise there is race condition between consume and add, and listener might be lost. it's also the same for current implementation, please see https://github.com/apache/helix/blob/master/zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java#L290

On the other hand, the use of HashMap instead of ConcurrentHashMap could work, but it requires we use a read lock on iteration read. In our use case, reads vastly out number writes, so a lock-free read is preferable.